File: backend/src/main/java/com/petconnect/backend/user/application/service/ClinicStaffService.java
package com.petconnect.backend.user.application.service;

import com.petconnect.backend.user.application.dto.ClinicStaffCreationDto;
import com.petconnect.backend.user.application.dto.ClinicStaffProfileDto;
import com.petconnect.backend.user.application.dto.ClinicStaffUpdateDto;
import jakarta.persistence.EntityExistsException;
import jakarta.persistence.EntityNotFoundException;
import org.springframework.lang.Nullable;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Service interface for managing Clinic Staff members (Vets and Admins).
 * Handles creation, activation/deactivation, and potentially updates initiated by Admins.
 *
 * @author ibosquet
 */
public interface ClinicStaffService {
    /**
     * Creates a new clinic staff member (VET or ADMIN) with the specified details.
     * This method is executed by an ADMIN and optionally uploads the public and
     * private key files for secure operations, typically required for VET roles.
     *
     * @param creationDTO A DTO containing the details of the staff member to be created,
     *                    such as username, email, name, role (VET or ADMIN), etc.
     * @param publicKeyFile The public key file for the staff member may be null if not applicable.
     * @param privateKeyFileEncrypted The encrypted private key file for the staff member may be null if not applicable.
     * @param creatingAdminId The unique ID of the ADMIN creating the staff member. Used for authorization and auditing.
     * @return A ClinicStaffProfileDto representing the newly created staff member's profile data.
     * @throws EntityExistsException if a user with the provided username or email already exists.
     * @throws AccessDeniedException if the creating admin lacks the necessary permissions.
     * @throws IllegalArgumentException if the provided data is invalid or missing required fields.
     */
    ClinicStaffProfileDto createClinicStaff(ClinicStaffCreationDto creationDTO,
                                            @Nullable MultipartFile publicKeyFile,
                                            @Nullable MultipartFile privateKeyFileEncrypted,
                                            Long creatingAdminId);

    /**
     * Activates a Clinic Staff member account.
     * Activated staff can log in and perform their designated roles.
     *
     * @param staffId The ID of the staff member to activate.
     * @param updatingAdminId The ID of the ADMIN performing the activation.
     * @return A ClinicStaffProfileDto representing the activated staff member's profile details.
     * @throws EntityNotFoundException if the specified staff member does not exist.
     * @throws AccessDeniedException if the updating user lacks necessary permissions.
     * @throws IllegalStateException if the staff member is already active.
     */
    ClinicStaffProfileDto activateStaff(Long staffId,
                                        Long updatingAdminId);

    /**
     * Deactivates a Clinic Staff member account.
     * Deactivated staff can no longer log in, but their records remain.
     *
     * @param staffId The ID of the staff member to deactivate.
     * @param deactivatingAdminId The ID of the ADMIN performing the action.
     * @return The profile DTO of the deactivated staff member.
     * @throws EntityNotFoundException if the staff member is not found.
     * @throws AccessDeniedException if the deactivating user is not authorized.
     * @throws IllegalStateException if the staff member is already inactive or if trying to deactivate oneself.
     */
    ClinicStaffProfileDto deactivateStaff(Long staffId, Long deactivatingAdminId);


    /**
     * Finds all active staff members for a given clinic.
     *
     * @param clinicId The ID of the clinic.
     * @return A list of ClinicStaffProfileDto for active staff.
     */
    List<ClinicStaffProfileDto> findActiveStaffByClinic(Long clinicId, Long requesterUserId);

    /**
     * Finds all staff members (active and inactive) for a given clinic.
     * Useful for administrative views.
     *
     * @param clinicId The ID of the clinic.
     * @return A list of ClinicStaffProfileDto for all staff.
     */
    List<ClinicStaffProfileDto> findAllStaffByClinic(Long clinicId, Long requesterUserId);

    /**
     * Updates the details of an existing Clinic Staff member.
     * This operation can only be performed by an ADMIN user with proper authorization.
     * The update includes editable fields in the ClinicStaffUpdateDto and optionally uploads a public key file
     * if the staff role is VET.
     *
     * @param staffId The unique ID of the staff member to update.
     * @param updateDTO A DTO containing updated staff details such as name, surname, and license number (if applicable).
     * @param publicKeyFile The public key file for the staff member's digital signature capability. May be null if not applicable.
     * @param updatingAdminId The ID of the ADMIN user performing the update operation.
     * @return A ClinicStaffProfileDto representing the updated staff member's profile details.
     * @throws EntityNotFoundException if the specified staff member does not exist.
     * @throws AccessDeniedException if the updating user lacks authorization.
     * @throws IllegalArgumentException if the provided update data is invalid or missing required fields.
     */
    ClinicStaffProfileDto updateClinicStaff(Long staffId, ClinicStaffUpdateDto updateDTO, @Nullable MultipartFile publicKeyFile, @Nullable MultipartFile privateKeyFile, Long updatingAdminId);

}

File: frontend/src/services/clinicStaffService.ts
import axios from "axios";
import { API_BASE_URL } from "../config";
import {
  ClinicStaffProfile,
  ApiErrorResponse,
  ClinicStaffUpdatePayload
} from "../types/apiTypes";

/**
 * Creates a new clinic staff member (Vet or Admin) for the admin's clinic.
 * Sends data as multipart/form-data if a public key file is included for a VET.
 *
 * @param {string} token - The JWT token of the authenticated Admin.
 * @param {FormData} formData - The FormData object containing 'dto' (JSON blob) and optionally 'publicKeyFile'.
 * @returns {Promise<ClinicStaffProfile>} A promise resolving to the profile of the newly created staff member.
 * @throws {Error} Throws an error if creation fails.
 */
export const createClinicStaff = async (
  token: string,
  formData: FormData
): Promise<ClinicStaffProfile> => {
  if (!token) {
    throw new Error("Authentication token required.");
  }

  try {
    const response = await axios.post<ClinicStaffProfile>(
      `${API_BASE_URL}/staff`,
      formData,
      {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      }
    );
    return response.data;
  } catch (error) {
    if (axios.isAxiosError(error) && error.response) {
      const apiError = error.response.data as ApiErrorResponse;
      console.error("API Create Staff Error:", apiError);
      let errorMessage = "Failed to create staff member.";
      if (typeof apiError.message === "string") {
        errorMessage = apiError.message;
      } else if (
        typeof apiError.message === "object" &&
        apiError.message !== null
      ) {
        errorMessage = Object.values(apiError.message).join(" ");
      } else if (apiError.error) {
        errorMessage = apiError.error;
      }
      if (error.response.status === 409)
        errorMessage = `Creation failed: ${errorMessage}`;
      throw new Error(errorMessage);
    } else {
      console.error("Network or unexpected create staff error:", error);
      throw new Error(
        "Failed to create staff member due to network or unexpected error."
      );
    }
  }
};

/**
 * Fetches all staff members (active and inactive) for a specific clinic.
 * Requires ADMIN or VET role of that clinic.
 *
 * @param {string} token - The JWT token.
 * @param {number | string} clinicId - The ID of the clinic.
 * @returns {Promise<ClinicStaffProfile[]>} A promise resolving to a list of staff profiles.
 * @throws {Error} Throws an error if fetching fails.
 */
export const getAllStaffForClinic = async (
  token: string,
  clinicId: number | string
): Promise<ClinicStaffProfile[]> => {
  if (!token) {
    throw new Error("Authentication token required.");
  }
  if (!clinicId) {
    throw new Error("Clinic ID is required.");
  }

  try {
    const response = await axios.get<ClinicStaffProfile[]>(
      `${API_BASE_URL}/clinics/${clinicId}/staff/all`,
      {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      }
    );
    return response.data;
  } catch (error) {
    if (axios.isAxiosError(error) && error.response) {
      const apiError = error.response.data as ApiErrorResponse;
      console.error("API Get All Staff Error:", apiError);
      throw new Error(
        typeof apiError.message === "string"
          ? apiError.message
          : apiError.error || "Failed to fetch staff list."
      );
    } else {
      console.error("Network or unexpected get all staff error:", error);
      throw new Error(
        "Failed to fetch staff list due to network or unexpected error."
      );
    }
  }
};

/**
 * Activates a clinic staff member's account.
 * Requires ADMIN role of that clinic.
 *
 * @param {string} token - The JWT token of the authenticated Admin.
 * @param {number | string} staffId - The ID of the staff member to activate.
 * @returns {Promise<ClinicStaffProfile>} A promise resolving to the profile of the activated staff member.
 * @throws {Error} Throws an error if activation fails.
 */
export const activateStaffMember = async (
  token: string,
  staffId: number | string
): Promise<ClinicStaffProfile> => {
  if (!token) {
    throw new Error("Authentication token required.");
  }

  try {
    const response = await axios.put<ClinicStaffProfile>(
      `${API_BASE_URL}/staff/${staffId}/activate`,
      {},
      {
        headers: { Authorization: `Bearer ${token}` },
      }
    );
    return response.data;
  } catch (error) {
    if (axios.isAxiosError(error) && error.response) {
      const apiError = error.response.data as ApiErrorResponse;
      console.error("API Activate Staff Error:", apiError);
      throw new Error(
        typeof apiError.message === "string"
          ? apiError.message
          : apiError.error || "Failed to activate staff."
      );
    } else {
      console.error("Network or unexpected activate staff error:", error);
      throw new Error(
        "Failed to activate staff due to network or unexpected error."
      );
    }
  }
};

/**
 * Deactivates a clinic staff member's account.
 * Requires ADMIN role of that clinic. Admin cannot deactivate themselves.
 *
 * @param {string} token - The JWT token of the authenticated Admin.
 * @param {number | string} staffId - The ID of the staff member to deactivate.
 * @returns {Promise<ClinicStaffProfile>} A promise resolving to the profile of the deactivated staff member.
 * @throws {Error} Throws an error if deactivation fails.
 */
export const deactivateStaffMember = async (
  token: string,
  staffId: number | string
): Promise<ClinicStaffProfile> => {
  if (!token) {
    throw new Error("Authentication token required.");
  }

  try {
    const response = await axios.put<ClinicStaffProfile>(
      `${API_BASE_URL}/staff/${staffId}/deactivate`,
      {},
      {
        // Empty body
        headers: { Authorization: `Bearer ${token}` },
      }
    );
    return response.data;
  } catch (error) {
    if (axios.isAxiosError(error) && error.response) {
      const apiError = error.response.data as ApiErrorResponse;
      console.error("API Deactivate Staff Error:", apiError);
      throw new Error(
        typeof apiError.message === "string"
          ? apiError.message
          : apiError.error || "Failed to deactivate staff."
      );
    } else {
      console.error("Network or unexpected deactivate staff error:", error);
      throw new Error(
        "Failed to deactivate staff due to network or unexpected error."
      );
    }
  }
};

/**
 * Updates an existing clinic staff member.
 * Handles sending data as multipart/form-data if a new public key file is provided.
 *
 * @param {string} token - The JWT token of the authenticated Admin.
 * @param {number | string} staffId - The ID of the staff member to update.
 * @param {ClinicStaffUpdatePayload} updateData - The data to update (JSON part).
 * @param {File | null} publicKeyFile - The optional new public key file.
 * @returns {Promise<ClinicStaffProfile>} A promise resolving to the updated staff profile.
 * @throws {Error} Throws an error if update fails.
 */
export const updateClinicStaff = async (
    token: string,
    staffId: number | string,
    updateData: ClinicStaffUpdatePayload,
    publicKeyFile: File | null,
    privateKeyFile: File | null
): Promise<ClinicStaffProfile> => {
    if (!token) { throw new Error("Authentication token required."); }

    const formData = new FormData();
    const dtoBlob = new Blob([JSON.stringify(updateData)], { type: 'application/json' });
    formData.append('dto', dtoBlob);

    if (publicKeyFile) {
        formData.append('publicKeyFile', publicKeyFile, publicKeyFile.name);
        console.log(`Updating staff ${staffId} with new public key file.`);
    } else {
        console.log(`Updating staff ${staffId} without changing public key file.`);
    }

    if (privateKeyFile) { 
        formData.append('privateKeyFile', privateKeyFile, privateKeyFile.name);
    } else {
        console.log(`Updating staff ${staffId} without changing private key file.`);
    }

    try {
        const response = await axios.put<ClinicStaffProfile>(`${API_BASE_URL}/staff/${staffId}`, formData, {
            headers: {
                'Authorization': `Bearer ${token}`,
            }
        });
        return response.data;
    } catch (error) {
         if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
            console.error(`API Update Staff (${staffId}) Error:`, apiError);
            let errorMessage = 'Failed to update staff member.';
             if (typeof apiError.message === 'string') { errorMessage = apiError.message; }
             else if (typeof apiError.message === 'object' && apiError.message !== null) { errorMessage = Object.values(apiError.message).join(' '); }
             else if (apiError.error) { errorMessage = apiError.error; }
             if (error.response.status === 409) errorMessage = `Update failed: ${errorMessage}`;
            throw new Error(errorMessage);
        } else {
            console.error(`Network or unexpected update staff (${staffId}) error:`, error);
            throw new Error('Failed to update staff member due to network or unexpected error.');
        }
    }
};
